### PR TITLE
Role management

### PR DESCRIPTION
**Pull request by geeksareforlife**
Most of this is complete now, needs a bit of extra work

### CODE REVIEW

File: app/HMS/User/Permissions/RoleManager.php [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Create a new RoleManager instance.
+     *
+     * @param HMS\Repositories\RoleRepository $roleRepository An instance of a role repository
+     */
+    public function __construct(RoleRepository $roleRepository)
+    {
+        $this->roleRepository = $roleRepository;
+        $this->permissionRepository = EntityManager::getRepository(Permission::class);
```
Comment (Hodgy) [inline app/HMS/User/Permissions/RoleManager.php]: Can we inject this instead of creating it here?

Comment (dpslwk) [inline app/HMS/User/Permissions/RoleManager.php]: I'm not sure we can inject generic repositories that are created on the fly

Comment (dpslwk) [inline app/HMS/User/Permissions/RoleManager.php]: looks like we would have to add it via the appServiceProvider and inject an ObjectRepository
not sure it thats going to obfuscate it away to much and if it used in various places we will never be 100% sure just which one is being injected

Comment (Hodgy) [inline app/HMS/User/Permissions/RoleManager.php]: Okay. In that case inject the entity manager and use that to create the repository is ready of the facade?

File: app/Http/Controllers/UserController.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * Show a specific user.
+     *
+     * @param int $id ID of the user
+     * @return \Illuminate\Http\Response
+     */
+    public function show($id)
+    {
+        $user = Auth::user();
+        if ( ! $user->hasPermissionTo('profile.view.all')) {
```
Comment (Hodgy) [inline app/Http/Controllers/UserController.php]: This should be moved into the constructor

File: routes/web.php [hunk 3]
```diff
@@ -30,6 +30,18 @@
 Route::get('register/{token}', 'Auth\RegisterController@showRegistrationForm')->name('register');
 Route::post('register', 'Auth\RegisterController@register');
 
+Route::group(['middleware' => 'auth'], function () {
+    // ROLE
+    Route::get('/roles', 'RoleController@index')->name('roles.index')->middleware('auth');
```
Comment (Hodgy) [inline routes/web.php]: Remove the middleware

File: app/HMS/User/Permissions/RoleManager.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace HMS\User\Permissions;
+
+use HMS\Repositories\RoleRepository;
+use Doctrine\ORM\EntityManagerInterface;
+use LaravelDoctrine\ACL\Permissions\Permission;
+
+class RoleManager
+{
+    private $roleRepository;
```
Comment (dpslwk) [inline app/HMS/User/Permissions/RoleManager.php]: missing Doc blocks

File: app/HMS/User/Permissions/RoleManager.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        $this->permissionRepository = $em->getRepository(Permission::class);
+    }
+
+    /**
+     * Update a specific role.
+     *
+     * @param Role $role the Role we want to update
+     * @param array $details fields that need updating
+     */
+    public function updateRole($role, $details)
```
Comment (dpslwk) [inline app/HMS/User/Permissions/RoleManager.php]: Type hint to Role and use

File: app/Http/Controllers/RoleController.php [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        $this->roleManager->updateRole($role, $request->all());
+
+        return redirect()->route('roles.show', ['role' => $role->getId()]);
+    }
+
+    /**
+     * Remove a specific user from a specific role.
+     *
+     * @param int $roleId ID of the role
+     * @param int $userId ID of the user
```
Comment (dpslwk) [inline app/Http/Controllers/RoleController.php]: Update doc block for type

**Pull request merged**